File: Ahorn/entities/Monumentbooster.jl
module EmHelperMonumentbooster

using ..Ahorn, Maple

@mapdef Entity "EmHelper/Monumentbooster" MonumentBooster(x::Integer, y::Integer, pattern::Integer=0, red::Bool=false, color::String="82d9ff", active::Bool=true)

const placements = Ahorn.PlacementDict(
    "Monument Booster (EmHelper, green)" => Ahorn.EntityPlacement(
        MonumentBooster,
        "point",
        Dict{String, Any}(
            "red" => false
        )
    ),
    "Monument Booster (EmHelper, red)" => Ahorn.EntityPlacement(
        MonumentBooster,
        "point",
        Dict{String, Any}(
            "red" => true
        )
    ),
)

function boosterSprite(entity::MonumentBooster)
    red = get(entity.data, "red", false)
    
    if red
        return "objects/monumentbooster/eboosterRed00"

    else
        return "objects/monumentbooster/ebooster00"
    end
end

function getColor(color)
    if haskey(Ahorn.XNAColors.colors, color)
        return Ahorn.XNAColors.colors[color]

    else
        try
            return ((Ahorn.argb32ToRGBATuple(parse(Int, replace(color, "#" => ""), base=16))[1:3] ./ 255)..., 1.0)

        catch

        end
    end

    return (1.0, 1.0, 1.0, 1.0)
end

function Ahorn.selection(entity::MonumentBooster)
    x, y = Ahorn.position(entity)
    sprite = boosterSprite(entity)

    return Ahorn.getSpriteRectangle(sprite, x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::MonumentBooster, room::Maple.Room)
    sprite = boosterSprite(entity)
    esprite = "objects/monumentbooster/e"
    entitycolor = String(get(entity.data, "color", "82d9ff"))
    entitytint = getColor(entitycolor)
    Ahorn.drawSprite(ctx, sprite, 0, 0)
    Ahorn.drawSprite(ctx, esprite, 0, 0, tint = entitytint)
end

end

File: Loenn/entities/Monumentbooster.lua
local drawableSprite = require("structs.drawable_sprite")
local xnaColors = require("consts.xna_colors")
local utils = require("utils")
local Monumentbooster = {}

Monumentbooster.name = "EmHelper/Monumentbooster"
Monumentbooster.depth = -8500

Monumentbooster.fieldInformation = {
    color = {
        fieldType = "color",
        allowXNAColors = true,
    }
}

Monumentbooster.placements = {
    {
        name = "Monument Booster (Green) [EmHelper]",
        data = {
            pattern = 0,
            red = false,
            color = "LightSkyBlue",
            active = true
        }
    },
    {
        name = "Monument Booster (Red) [EmHelper]",
        data = {
            pattern = 0,
            red = true,
            color = "LightSkyBlue",
            active = true
        }
    }
}

local eTexture = "objects/monumentbooster/e"
local bubbleTexture = "objects/monumentbooster/ebooster00"

function Monumentbooster.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local red = entity.red

    if red then
        bubbleTexture = "objects/monumentbooster/eboosterRed00"
    else
        bubbleTexture = "objects/monumentbooster/ebooster00"
    end

    local bubbleSprite = drawableSprite.fromTexture(bubbleTexture, {x = x, y = y})
    local eSprite = drawableSprite.fromTexture(eTexture, entity)

    bubbleSprite:setJustification(1, 1)
    bubbleSprite:addPosition(16, 16)

    eSprite:setJustification(1, 1)
    eSprite:addPosition(16, 16)

    return {
        bubbleSprite, 
        eSprite,
    }
end

function Monumentbooster.rectangle(room, entity)
    return utils.rectangle(entity.x-8, entity.y-8, 16, 16)
end

return Monumentbooster